File: ios/App/App/Plugins/BroadcastRemoteIdPlugin/OpenDroneIdParser.swift
import Foundation

/// Parser for ASTM F3411 / ASD-STAN prEN 4709-002 OpenDroneID BLE messages.
///
/// Each message is 25 bytes. The header byte encodes `(type << 4 | protoVersion)`.
/// Service data from BLE advertisement UUID 0xFFFA may contain a single message
/// or a message pack (type 0xF) with multiple 25-byte messages.
struct OpenDroneIdParser {

    // MARK: - Message Types

    private static let msgBasicId    = 0
    private static let msgLocation   = 1
    private static let msgAuth       = 2
    private static let msgSelfId     = 3
    private static let msgSystem     = 4
    private static let msgOperatorId = 5
    private static let msgPack       = 0xF

    private static let msgSize = 25

    // MARK: - Result

    struct DroneInfo {
        var uasId: String = ""
        var idType: Int = 0
        var uaType: Int = 0
        var latitude: Double = 0
        var longitude: Double = 0
        var altGeo: Double = -1000
        var speed: Double = 0
        var heading: Double = 0
        var vertSpeed: Double = 0
        var status: Int = 0
        var operatorLat: Double = 0
        var operatorLon: Double = 0
        var operatorRegistrationId: String = ""
        var description: String = ""
        var timestamp: Int = 0
    }

    // MARK: - Public API

    /// Parse BLE service data for UUID 0xFFFA.
    /// The first byte is an application code (counter); actual messages start at offset 1.
    static func parseServiceData(_ data: Data) -> DroneInfo {
        var info = DroneInfo()
        guard data.count >= 2 else { return info }

        let offset = 1 // skip application code byte
        let remaining = data.count - offset
        guard remaining >= msgSize else { return info }

        let header = Int(data[offset])
        let msgType = (header >> 4) & 0x0F

        if msgType == msgPack {
            guard remaining >= 2 else { return info }
            let msgCount = Int(data[offset + 1])
            var packOffset = offset + 2
            for _ in 0..<msgCount {
                guard packOffset + msgSize <= data.count else { break }
                parseSingleMessage(data, offset: packOffset, info: &info)
                packOffset += msgSize
            }
        } else {
            parseSingleMessage(data, offset: offset, info: &info)
        }

        return info
    }

    // MARK: - Private

    private static func parseSingleMessage(_ data: Data, offset: Int, info: inout DroneInfo) {
        guard offset + msgSize <= data.count else { return }

        let header = Int(data[offset])
        let msgType = (header >> 4) & 0x0F
        let p = offset + 1 // payload start

        switch msgType {
        case msgBasicId:
            parseBasicId(data, p: p, info: &info)
        case msgLocation:
            parseLocation(data, p: p, info: &info)
        case msgSelfId:
            parseSelfId(data, p: p, info: &info)
        case msgSystem:
            parseSystem(data, p: p, info: &info)
        case msgOperatorId:
            parseOperatorId(data, p: p, info: &info)
        default:
            break // auth and unknown types ignored
        }
    }

    private static func parseBasicId(_ data: Data, p: Int, info: inout DroneInfo) {
        info.idType = (Int(data[p]) & 0xF0) >> 4
        info.uaType = Int(data[p]) & 0x0F
        info.uasId = extractString(data, offset: p + 1, maxLen: 20)
    }

    private static func parseLocation(_ data: Data, p: Int, info: inout DroneInfo) {
        let flags = Int(data[p])
        info.status = (flags >> 4) & 0x0F
        let speedMult = flags & 0x03
        let speedMultiplier: Double = speedMult == 0 ? 0.25 : (speedMult == 1 ? 0.75 : 1.0)

        let dirRaw = Int(data[p + 1])
        let ewBit = ((flags >> 2) & 0x01) == 1
        info.heading = Double((dirRaw + (ewBit ? 180 : 0)) % 360)

        let speedRaw = Int(data[p + 2])
        info.speed = Double(speedRaw) * speedMultiplier

        info.vertSpeed = Double(Int8(bitPattern: data[p + 3])) * 0.5

        info.latitude = Double(readInt32LE(data, offset: p + 4)) / 1e7
        info.longitude = Double(readInt32LE(data, offset: p + 8)) / 1e7

        let altGeoRaw = Int(readUInt16LE(data, offset: p + 14))
        info.altGeo = Double(altGeoRaw) * 0.5 - 1000

        info.timestamp = Int(readUInt16LE(data, offset: p + 22))
    }

    private static func parseSelfId(_ data: Data, p: Int, info: inout DroneInfo) {
        info.description = extractString(data, offset: p + 1, maxLen: 23)
    }

    private static func parseSystem(_ data: Data, p: Int, info: inout DroneInfo) {
        info.operatorLat = Double(readInt32LE(data, offset: p + 1)) / 1e7
        info.operatorLon = Double(readInt32LE(data, offset: p + 5)) / 1e7
    }

    private static func parseOperatorId(_ data: Data, p: Int, info: inout DroneInfo) {
        info.operatorRegistrationId = extractString(data, offset: p + 1, maxLen: 20)
    }

    // MARK: - Byte Helpers

    private static func readInt32LE(_ data: Data, offset: Int) -> Int32 {
        guard offset + 4 <= data.count else { return 0 }
        return data.withUnsafeBytes { ptr in
            ptr.loadUnaligned(fromByteOffset: offset, as: Int32.self)
        }
    }

    private static func readUInt16LE(_ data: Data, offset: Int) -> UInt16 {
        guard offset + 2 <= data.count else { return 0 }
        return data.withUnsafeBytes { ptr in
            ptr.loadUnaligned(fromByteOffset: offset, as: UInt16.self)
        }
    }

    private static func extractString(_ data: Data, offset: Int, maxLen: Int) -> String {
        let end = min(offset + maxLen, data.count)
        guard offset < end else { return "" }
        var bytes: [UInt8] = []
        for i in offset..<end {
            let b = data[i]
            if b == 0 { break }
            bytes.append(b)
        }
        return (String(bytes: bytes, encoding: .ascii) ?? "").trimmingCharacters(in: .whitespaces)
    }
}

File: android/app/src/main/java/ae/trustsky/spotlight/plugins/OpenDroneIdParser.java
package ae.trustsky.spotlight.plugins;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;

/**
 * Parser for ASTM F3411 / ASD-STAN prEN 4709-002 OpenDroneID BLE messages.
 *
 * Each message is 25 bytes. The header byte encodes (type << 4 | protoVersion).
 * Service data from BLE advertisement UUID 0xFFFA may contain a single message
 * (25 bytes) or a message pack (type 0xF) containing multiple 25-byte messages.
 */
public class OpenDroneIdParser {

    // Message types
    private static final int MSG_BASIC_ID     = 0;
    private static final int MSG_LOCATION     = 1;
    private static final int MSG_AUTH         = 2;
    private static final int MSG_SELF_ID      = 3;
    private static final int MSG_SYSTEM       = 4;
    private static final int MSG_OPERATOR_ID  = 5;
    private static final int MSG_PACK         = 0xF;

    private static final int MSG_SIZE = 25;

    private OpenDroneIdParser() {
        // Utility class — prevent instantiation
    }

    /** Accumulated result from parsing one or more messages. */
    public static class DroneInfo {
        private String uasId = "";
        private int idType = 0;
        private int uaType = 0;
        private double latitude = 0;
        private double longitude = 0;
        private double altGeo = -1000;
        private double speed = 0;
        private double heading = 0;
        private double vertSpeed = 0;
        private int status = 0;
        private double operatorLat = 0;
        private double operatorLon = 0;
        private String operatorRegistrationId = "";
        private String description = "";
        private long timestamp = 0;

        public String getUasId() { return uasId; }
        public int getIdType() { return idType; }
        public int getUaType() { return uaType; }
        public double getLatitude() { return latitude; }
        public double getLongitude() { return longitude; }
        public double getAltGeo() { return altGeo; }
        public double getSpeed() { return speed; }
        public double getHeading() { return heading; }
        public double getVertSpeed() { return vertSpeed; }
        public int getStatus() { return status; }
        public double getOperatorLat() { return operatorLat; }
        public double getOperatorLon() { return operatorLon; }
        public String getOperatorRegistrationId() { return operatorRegistrationId; }
        public String getDescription() { return description; }
        public long getTimestamp() { return timestamp; }
    }

    /**
     * Parse BLE service data for UUID 0xFFFA.
     * After Android strips the 16-bit UUID, the data layout is:
     *   Byte 0: Application Code (0x0D for OpenDroneID)
     *   Byte 1: 8-bit message counter
     *   Bytes 2+: 25-byte OpenDroneID message(s)
     */
    public static DroneInfo parseServiceData(byte[] data) {
        DroneInfo info = new DroneInfo();
        if (data == null || data.length < 3) return info;

        // Skip application code (0x0D) and message counter
        int offset = 2;
        int remaining = data.length - offset;

        if (remaining < MSG_SIZE) return info;

        int header = data[offset] & 0xFF;
        int msgType = (header >> 4) & 0x0F;

        if (msgType == MSG_PACK) {
            // Message pack: header byte, then 1 byte msgCount, then N * 25 messages
            if (remaining < 2) return info;
            int msgCount = data[offset + 1] & 0xFF;
            int packOffset = offset + 2;
            for (int i = 0; i < msgCount && packOffset + MSG_SIZE <= data.length; i++) {
                parseSingleMessage(data, packOffset, info);
                packOffset += MSG_SIZE;
            }
        } else {
            parseSingleMessage(data, offset, info);
        }

        return info;
    }

    private static void parseSingleMessage(byte[] data, int offset, DroneInfo info) {
        if (offset + MSG_SIZE > data.length) return;

        int header = data[offset] & 0xFF;
        int msgType = (header >> 4) & 0x0F;

        // Payload starts after the header byte
        int p = offset + 1;

        switch (msgType) {
            case MSG_BASIC_ID:
                parseBasicId(data, p, info);
                break;
            case MSG_LOCATION:
                parseLocation(data, p, info);
                break;
            case MSG_SELF_ID:
                parseSelfId(data, p, info);
                break;
            case MSG_SYSTEM:
                parseSystem(data, p, info);
                break;
            case MSG_OPERATOR_ID:
                parseOperatorId(data, p, info);
                break;
            case MSG_AUTH:
                // Authentication messages are optional; skip for now
                break;
            default:
                // Unknown message type; ignore
                break;
        }
    }

    private static void parseBasicId(byte[] data, int p, DroneInfo info) {
        // Byte 0: idType (4 bits) | uaType (4 bits)
        info.idType = (data[p] & 0xF0) >> 4;
        info.uaType = data[p] & 0x0F;
        // Bytes 1..20: UAS ID (null-terminated ASCII)
        info.uasId = extractString(data, p + 1, 20);
    }

    private static void parseLocation(byte[] data, int p, DroneInfo info) {
        ByteBuffer bb = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN);

        // Byte 0: status (4 bits) | heightType (1 bit) | EW direction (1 bit) | speedMult (2 bits)
        int flags = data[p] & 0xFF;
        info.status = (flags >> 4) & 0x0F;
        int speedMult = flags & 0x03;
        double speedMultiplier = resolveSpeedMultiplier(speedMult);

        // Byte 1: direction — EW bit selects 0-179 or 180-359 range
        int dirRaw = data[p + 1] & 0xFF;
        boolean ewBit = ((flags >> 2) & 0x01) == 1;
        info.heading = (dirRaw + (ewBit ? 180 : 0)) % 360;

        // Byte 2: speed (encoded, depends on multiplier)
        int speedRaw = data[p + 2] & 0xFF;
        info.speed = speedRaw * speedMultiplier;

        // Byte 3: vertical speed (int8, resolution 0.5 m/s)
        info.vertSpeed = data[p + 3] * 0.5;

        // Bytes 4-7: latitude (int32 LE, degrees * 1e7)
        info.latitude = bb.getInt(p + 4) / 1e7;

        // Bytes 8-11: longitude (int32 LE, degrees * 1e7)
        info.longitude = bb.getInt(p + 8) / 1e7;

        // Bytes 14-15: geodetic altitude (uint16 LE, * 0.5 - 1000)
        int altGeoRaw = bb.getShort(p + 14) & 0xFFFF;
        info.altGeo = altGeoRaw * 0.5 - 1000;

        // Bytes 22-23: timestamp (uint16 LE, tenths of seconds since the hour)
        int tsRaw = bb.getShort(p + 22) & 0xFFFF;
        info.timestamp = tsRaw;
    }

    private static double resolveSpeedMultiplier(int speedMult) {
        if (speedMult == 0) return 0.25;
        if (speedMult == 1) return 0.75;
        return 1.0;
    }

    private static void parseSelfId(byte[] data, int p, DroneInfo info) {
        // Byte 0: description type
        // Bytes 1..23: description text (null-terminated)
        info.description = extractString(data, p + 1, 23);
    }

    private static void parseSystem(byte[] data, int p, DroneInfo info) {
        ByteBuffer bb = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN);

        // Bytes 1-4: operator latitude (int32 LE / 1e7)
        info.operatorLat = bb.getInt(p + 1) / 1e7;
        // Bytes 5-8: operator longitude (int32 LE / 1e7)
        info.operatorLon = bb.getInt(p + 5) / 1e7;
    }

    private static void parseOperatorId(byte[] data, int p, DroneInfo info) {
        // Bytes 1..20: operator registration ID (null-terminated ASCII)
        info.operatorRegistrationId = extractString(data, p + 1, 20);
    }

    /**
     * Find the OpenDroneID payload in raw BLE advertisement bytes.
     * Scans for AD type 0x16 with UUID 0xFFFA and app code 0x0D.
     * Returns the index of the app code byte, or -1 if not found.
     */
    public static int findOdidPayload(byte[] raw) {
        if (raw == null) return -1;
        int i = 0;
        while (i < raw.length) {
            int adLen = raw[i] & 0xFF;
            if (adLen == 0 || i + adLen >= raw.length) break;
            // AD type at i+1, data starts at i+2
            int adType = raw[i + 1] & 0xFF;
            if (adType == 0x16 && adLen >= 5) {
                // Check for UUID 0xFFFA (little-endian: FA FF)
                int uuidLo = raw[i + 2] & 0xFF;
                int uuidHi = raw[i + 3] & 0xFF;
                if (uuidLo == 0xFA && uuidHi == 0xFF) {
                    // App code at i+4, counter at i+5, message at i+6
                    return i + 4;
                }
            }
            i += adLen + 1;
        }
        return -1;
    }

    /** Extract a null-terminated ASCII string from a byte array. */
    private static String extractString(byte[] data, int offset, int maxLen) {
        int end = Math.min(offset + maxLen, data.length);
        StringBuilder sb = new StringBuilder();
        for (int i = offset; i < end; i++) {
            char c = (char) (data[i] & 0xFF);
            if (c == 0) break;
            sb.append(c);
        }
        return sb.toString().trim();
    }
}
